### PR TITLE
Project reborn 

### PR DESCRIPTION
**Pull request by Aliaksandr-Yanushkevich**
Hello.
After fail with design of database, I've designed new database scheme and rewritten almost all components. Could you review my code?

Deploy link https://dims.netlify.app/members

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -4,30 +4,22 @@ import Header from './components/common/Header/Header';
...
+import Login from './components/Login/Login';
+import Registration from './components/Registration/Registration';
 
 class App extends Component {
   state = {
-    members: null,
-    currentUserId: null,
+    // currentUserId: 'newMember',
+    currentUserId: 'pu7f8c6lk',
+    currentTaskId: 'newTask',
```
Comment (maximsan) [inline src/App.js:18]: why do you need hard-code in the state ? use default props if you need

File: src/App.js [hunk 2]
```diff
@@ -37,41 +29,51 @@ class App extends Component {
...
-        this.setState({
-          members,
-        }),
-      )
-      .catch((error) => {
-        console.error(`Error receiving data: ${error}`);
-      });
+  setCurrentTask = (e) => {
+    e.persist();
+    const taskId = e.target.dataset.taskid;
```
Comment (maximsan) [inline src/App.js:34]: use spread operator and rename to currentTaskId to use short object props

File: src/api/firebaseApi.js [hunk 3]
```diff
@@ -1,74 +1,321 @@
...
-};
 
 const firebaseApi = {
-  getMembers() {
+  createUser(userId, userInfo) {
+    return firestore
+      .collection('UserProfile')
+      .doc(userId)
+      .set({
+        ...userInfo,
```
Comment (maximsan) [inline src/api/firebaseApi.js:11]: why do you spread it?

File: src/api/firebaseApi.js [hunk 4]
```diff
@@ -1,74 +1,321 @@
...
+      .get()
+      .then((users) => {
+        users.forEach((user) => {
+          const { firstName, lastName, userId } = user.data();
+          members = [
+            ...members,
+            {
+              firstName,
+              lastName,
+              userId,
```
Comment (maximsan) [inline src/api/firebaseApi.js:89]: why do you spread members an do not spread member ?

File: src/components/FormField/FormField.jsx [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+      }
+
+      if (!this.checkRequirement(minLength <= value.length, `Min length should be ${minLength} symbols`)) {
+        return;
+      }
+
+      if (!this.checkRequirement(maxLength >= value.length, `Max length should be ${maxLength} symbols`)) {
+      }
+    }
+  };
```
Comment (maximsan) [inline src/components/FormField/FormField.jsx:75]: extract validate and split in several helpers

File: src/components/FormField/FormField.jsx [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+              onFocus={this.onFocus}
+              onBlur={this.validate}
+              required={required}
+              className={message && touched ? styles.error : null}
+              placeholder={placeholder}
+            />
+          </div>
+          <p className={styles.message}>{message}</p>
+        </div>
+      );
```
Comment (maximsan) [inline src/components/FormField/FormField.jsx:115]: create component

File: src/components/FormField/FormField.jsx [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+            className={message && touched ? styles.error : null}
+            placeholder={placeholder}
+            min={min}
+            max={max}
+            step={step}
+            required={required}
+          />
+        </div>
+        <p className={styles.message}>{message}</p>
+      </div>
```
Comment (maximsan) [inline src/components/FormField/FormField.jsx:137]: create omponent

File: src/components/Login/Login.jsx [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    formIsValid: false,
+  };
+
+  login = () => {
+    const { login, password } = this.state;
+    firebaseApi.login(login, password);
+  };
+
+  onChange = (e) => {
+    debugger;
```
Comment (maximsan) [inline src/components/Login/Login.jsx:23]: remove

File: src/components/Login/Login.jsx [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    const { id, value, checked } = e.currentTarget;
+    id === 'remember' ? this.setState({ [id]: checked }) : this.setState({ [id]: value });
```
Comment (maximsan) [inline src/components/Login/Login.jsx:25]: rewrite to `if else`

File: src/components/MemberPage/MemberPage.jsx [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+      directionId: 0,
+      address: '',
+      education: '',
+      mathScore: '',
+      universityAverageScore: '',
+      formIsValid: false,
+      isFetching: false,
+    };
+    this.root = document.createElement('div');
+    document.body.appendChild(this.root);
```
Comment (maximsan) [inline src/components/MemberPage/MemberPage.jsx:36]: why do not use React.createRef() ?

File: src/components/MemberPage/MemberPage.jsx [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+            email,
+            startDate,
+            skype,
+            birthDate,
+            directionId,
+            address,
+            education,
+            mathScore,
+            universityAverageScore,
+          } = userInfo.data();
```
Comment (maximsan) [inline src/components/MemberPage/MemberPage.jsx:64]: move to your firebase helper `.getUserInfo(userId)`,  `userInfo.data()` and etc

File: src/components/MemberPage/MemberPage.jsx [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+      address,
+      education,
+      mathScore,
+      universityAverageScore,
+    };
+
+    firebaseApi.createUser(userId, userInfo).catch((error) => {
+      console.error(`User creation error: ${error}`);
+    });
+  };
```
Comment (maximsan) [inline src/components/MemberPage/MemberPage.jsx:149]: it can be separated helper

File: src/components/MemberPage/MemberPage.jsx [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+      mathScore >= 0 &&
+      mathScore <= 100 &&
+      universityAverageScore >= 0 &&
+      universityAverageScore <= 10
+    ) {
+      this.setState({ formIsValid: true });
+    } else {
+      this.setState({ formIsValid: false });
+    }
+  };
```
Comment (maximsan) [inline src/components/MemberPage/MemberPage.jsx:190]: it an be separated helpers, do not pollute you components with additional logic, it's bad

File: src/components/MemberPage/MemberPage.jsx [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+
+  onChange = (e) => {
+    const { id, value } = e.currentTarget;
+    let preparedValue = value;
```
Comment (maximsan) [inline src/components/MemberPage/MemberPage.jsx:194]: you cane rename value in one line

File: src/components/MemberPage/MemberPage.jsx [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+      isFetching,
+      directions,
+      formIsValid,
+    } = this.state;
+
+    const preparedDIrections = directions ? directionsToOptions(directions) : '';
+    const preparedGenders = [
+      { value: 'male', title: 'Male' },
+      { value: 'female', title: 'Female' },
+    ];
```
Comment (maximsan) [inline src/components/MemberPage/MemberPage.jsx:227]: Move it out of component

File: src/components/MemberTasks/MemberCurrentTasks.jsx [hunk 16]
```diff
@@ -1,40 +1,68 @@
...
+  const succesedTask = (e) => {
+    e.persist();
+    const currentTaskId = e.target.dataset.taskid;
+    firebaseApi.completeTask(currentTaskId, 'success');
+  };
+
+  const failedTask = (e) => {
+    e.persist();
+    const currentTaskId = e.target.dataset.taskid;
+    firebaseApi.completeTask(currentTaskId, 'fail');
```
Comment (maximsan) [inline src/components/MemberTasks/MemberCurrentTasks.jsx:28]: use the same method, just pass different `flag`

File: src/components/Members/MemberData.jsx [hunk 17]
```diff
@@ -1,59 +1,114 @@
...
+  };
+
+  componentDidMount() {
+    const directions = [];
+    firebaseApi
+      .getDirections()
+      .then((courseDirections) => {
+        courseDirections.forEach((direction) => {
+          const { directionId, name } = direction.data();
+          directions.push({ directionId, name });
```
Comment (maximsan) [inline src/components/Members/MemberData.jsx:23]: try to get used to not use mutate data methods in components

File: src/components/Members/MemberData.jsx [hunk 18]
```diff
@@ -1,59 +1,114 @@
...
+      deleteUser,
+      createUser,
+    } = this.props;
+    const { directions } = this.state;
+
+    const age = getAge(birthDate);
+    const direction = directions
+      ? directions.filter((courseDirection) => {
+          return courseDirection.directionId === directionId;
+        })[0].name
```
Comment (maximsan) [inline src/components/Members/MemberData.jsx:54]: use find

File: src/components/Registration/Registration.jsx [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+  };
+
+  register = () => {
+    const { email, password } = this.state;
+    firebaseApi.register(email, password);
+  };
+
+  onChange = (e) => {
+    const { id, value, checked } = e.currentTarget;
+    id === 'remember' ? this.setState({ [`${id}`]: checked }) : this.setState({ [`${id}`]: value });
```
Comment (maximsan) [inline src/components/Registration/Registration.jsx:24]: if else

File: src/components/Registration/Registration.jsx [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+            maxLength={30}
+            inputType='password'
+            id='repeatedPassword'
+            label='Repeate password'
+            value={repeatedPassword}
+            onChange={this.onChange}
+            validateForm={this.validateForm}
+          />
+          <div className={styles.item}>
+            {/* <Button disabled={!(loginIsValid && passwordIsValid && paswordsMatch)}>Register</Button> */}
```
Comment (maximsan) [inline src/components/Registration/Registration.jsx:66]: remove

File: src/components/TaskPage/TaskPage.jsx [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+      members: null,
+      name: '',
+      description: '',
+      startDate: dateToStringForInput(new Date()),
+      deadlineDate: '',
+      userTasks: [],
+      formIsValid: false,
+    };
+    this.root = document.createElement('div');
+    document.body.appendChild(this.root);
```
Comment (maximsan) [inline src/components/TaskPage/TaskPage.jsx:25]: react.createRef

File: src/components/common/Header/Header.jsx [hunk 22]
```diff
@@ -3,13 +3,25 @@ import { NavLink } from 'react-router-dom';
...
   return (
     <header className={styles.header}>
-      <NavLink to='/members'>
+      <NavLink className={styles.link} to='/members'>
         <img src={logo} alt='logo' />
       </NavLink>
+      <div className='nav'>
+        <NavLink className={styles.link} to='/members'>
+          <Button>Manage members</Button>
+        </NavLink>
```
Comment (maximsan) [inline src/components/common/Header/Header.jsx:17]: create NavButton component

File: src/helpers/directionsToOptions.js [hunk 23]
```diff
@@ -0,0 +1,12 @@
+const directionsToOptions = (directions) => {
+  const options = [];
+  directions.forEach(({ directionId, name }) =>
+    options.push({
+      value: directionId,
+      title: name,
+    }),
+  );
+  return options;
```
Comment (maximsan) [inline src/helpers/directionsToOptions.js:9]: you can use just map and return result at the same time

File: src/helpers/getGender.js [hunk 24]
```diff
@@ -0,0 +1,8 @@
+import faker from 'faker';
+
+const getGender = () => {
+  const genders = ['female', 'male'];
+  return faker.random.arrayElement(genders);
```
Comment (maximsan) [inline src/helpers/getGender.js:5]: you don't need to use whole library for this

File: src/App.js (hunk 1 above)
Comment (Aliaksandr-Yanushkevich) [reply src/App.js:18]: fixed

File: src/App.js (hunk 2 above)
Comment (Aliaksandr-Yanushkevich) [reply src/App.js:34]: fixed

File: src/api/firebaseApi.js (hunk 3 above)
Comment (Aliaksandr-Yanushkevich) [reply src/api/firebaseApi.js:11]: fixed

File: src/api/firebaseApi.js (hunk 4 above)
Comment (Aliaksandr-Yanushkevich) [reply src/api/firebaseApi.js:89]: fixed. used map instead foreach

File: src/components/FormField/FormField.jsx (hunk 5 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/FormField/FormField.jsx:75]: splited component on more specific (single responsibility principe). fixed

File: src/components/FormField/FormField.jsx (hunk 6 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/FormField/FormField.jsx:115]: fixed

File: src/components/FormField/FormField.jsx (hunk 7 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/FormField/FormField.jsx:137]: fixed

File: src/components/Login/Login.jsx (hunk 8 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/Login/Login.jsx:23]: fixed

File: src/components/Login/Login.jsx (hunk 9 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/Login/Login.jsx:25]: fixed

File: src/components/MemberPage/MemberPage.jsx (hunk 11 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/MemberPage/MemberPage.jsx:64]: fixed

File: src/components/MemberPage/MemberPage.jsx (hunk 12 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/MemberPage/MemberPage.jsx:149]: fixed

File: src/components/MemberPage/MemberPage.jsx (hunk 13 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/MemberPage/MemberPage.jsx:190]: fixed

File: src/components/MemberPage/MemberPage.jsx (hunk 15 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/MemberPage/MemberPage.jsx:227]: fixed

File: src/components/MemberTasks/MemberCurrentTasks.jsx (hunk 16 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/MemberTasks/MemberCurrentTasks.jsx:28]: fixed. I used innerText attribute and passed it to function

File: src/components/Members/MemberData.jsx (hunk 17 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/Members/MemberData.jsx:23]: fixed

File: src/components/Members/MemberData.jsx (hunk 18 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/Members/MemberData.jsx:54]: fixed

File: src/components/Registration/Registration.jsx (hunk 19 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/Registration/Registration.jsx:24]: this component had been deleted. I changed application logic. fixed

File: src/components/Registration/Registration.jsx (hunk 20 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/Registration/Registration.jsx:66]: this component had been deleted. I changed application logic. fixed

File: src/helpers/directionsToOptions.js (hunk 23 above)
Comment (Aliaksandr-Yanushkevich) [reply src/helpers/directionsToOptions.js:9]: fixed

File: src/helpers/getGender.js (hunk 24 above)
Comment (Aliaksandr-Yanushkevich) [reply src/helpers/getGender.js:5]: create simple custom randomaizer. fixed

File: src/components/common/Header/Header.jsx (hunk 22 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/common/Header/Header.jsx:17]: fixed

File: src/App.js (hunk 1 above)
Comment (Aliaksandr-Yanushkevich) [reply src/App.js:18]: fixed

File: src/components/MemberPage/MemberPage.jsx (hunk 10 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/MemberPage/MemberPage.jsx:36]: in the last PR I use Modal component for this purpose from reactstrap library, but I use useRef feature in MemberData component instead accessing to DOM.

File: src/components/TaskPage/TaskPage.jsx (hunk 21 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/TaskPage/TaskPage.jsx:25]: in the last PR I use Modal component for this purpose from reactstrap library, but I use useRef feature in MemberData component instead accessing to DOM.

File: src/components/MemberPage/MemberPage.jsx (hunk 14 above)
Comment (Aliaksandr-Yanushkevich) [reply src/components/MemberPage/MemberPage.jsx:194]: rewritten. fixed

**Comment by Aliaksandr-Yanushkevich** [comment]
I apologize for such a big delay in fixing bugs. I've corrected all these bugs in reactstrap PR

**Pull request merged**